src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -161,66 +260,80 @@ public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], o
 
         public async Task<ChannelReader<object>> StreamAsChannelAsync(string methodName, Type returnType, object[] args, CancellationToken cancellationToken = default)
         {
+            CheckDisposed();
+
             return await StreamAsChannelAsyncCore(methodName, returnType, args, cancellationToken).ForceAsync();
         }
 
         private async Task<ChannelReader<object>> StreamAsChannelAsyncCore(string methodName, Type returnType, object[] args, CancellationToken cancellationToken)
         {
-            if (!_startCalled)
+            await _connectionLock.WaitAsync();
```

Comment:
Does this mean only one thread can be inside stream/invoke/etc at a time? Won't this negatively effect performance?

Reply:
We already serialize writes further down. Pipelines are single-producer (they aren't thread-safe). But because of how Pipelines work, this will only go async when the server is applying backpressure to us (i.e. our local buffers are full), so we want to reduce the throughput there anyway.